app/utils/SQL.scala

```diff
@@ -0,0 +1,251 @@
+            outputValues ++= x.values
+          }
+          case x => {
+            val sqlValue = SqlValue.makeSqlValue(x)
+            outputSql ++= sqlValue.getPlaceholder()
+            outputValues += sqlValue
+          }
+        }
+      }
+    }
+
+    SqlToken(sql = outputSql.toString, values = outputValues.toList)
+  }
+}
+
+object NestedSQLInterpolation {
+  implicit def nestedSQLInterpolation(s: StringContext): NestedSQLInterpolator = new NestedSQLInterpolator(s)
+}
+
+case class SqlToken(sql: String, values: List[SqlValue] = List()) {
+  def debug: String = {
+    val parts = sql.split("\\?", -1)
+    assert(parts.tail.length == values.length)
+    parts.tail.zip(values).foldLeft(parts.head)((acc, x) => acc + x._2.debug() + x._1)
+  }
+
+  def as[R](implicit rconv: GetResult[R]): SqlStreamingAction[Vector[R], R, Effect] =
+    new StreamingInvokerAction[Vector[R], R, Effect] {
+      def statements = List(sql)
+
+      protected[this] def createInvoker(statements: Iterable[String]) = new StatementInvoker[R] {
+        val getStatement = statements.head
+
+        protected def setParam(st: PreparedStatement) = {
+          val pp = new PositionedParameters(st);
+          values.foreach(_.setParameter(pp))
+        }
+
+        protected def extractValue(rs: PositionedResult): R = rconv(rs)
+      }
+
+      override def getDumpInfo = DumpInfo(DumpInfo.simpleNameFor(getClass), mainInfo = s"[${debug}]")
+
+      protected[this] def createBuilder = Vector.newBuilder[R]
+    }
+
+  def asUpdate = as[Int](GetUpdateValue).head
+
+  def stripMargin(marginChar: Char): SqlToken = copy(sql = sql.stripMargin(marginChar))
+  def stripMargin: SqlToken = copy(sql = sql.stripMargin)
+}
+
+object SqlToken {
+  def join(values: List[Either[SqlValue, SqlToken]], sep: String): SqlToken = {
+    val outputSql = mutable.StringBuilder.newBuilder
+    val outputValues = ListBuffer[SqlValue]()
+    for (i <- values.indices) {
+      val value = values(i);
+      value match {
+        case Left(x) => {
+          outputSql ++= x.getPlaceholder()
+          outputValues += x
+        }
+        case Right(x) => {
+          outputSql ++= x.sql
+          outputValues ++= x.values
+        }
+      }
+      if (i < values.length - 1) {
+        outputSql ++= sep;
+      }
+    }
+    SqlToken(sql = outputSql.toString, values = outputValues.toList)
+  }
+
+  def tuple(values: List[Any]): SqlToken = {
+    val sqlValues = values.map(SqlValue.makeSqlValue)
+    SqlToken(sql = s"(${sqlValues.map(_.getPlaceholder()).mkString(", ")})", values = sqlValues)
+  }
+
+  def tupleList(values: List[List[Any]]): SqlToken = {
+    val sqlValueLists = values.map(list => list.map(SqlValue.makeSqlValue))
+    SqlToken(sql = sqlValueLists.map(list => s"(${list.map(_.getPlaceholder()).mkString(", ")})").mkString(", "),
+             values = sqlValueLists.flatten)
+  }
+
+  def raw(s: String) = SqlToken(s)
+
+  def empty(): SqlToken = raw("")
```

Comment:
```suggestion
  def empty: SqlToken = raw("")
```
(parens are usually added only where side effects happen)